### PR TITLE
avoid using mps for complex numbers

### LINKED ISSUE #432: Speeding up the process on Apple Silicon M1 Max
I'm running this on an M1 Max 64 GB and it's only running at 2 seconds. Is there any way to boost this?

### LINKED ISSUE #435: macOS Apple Silicon Metal GPU Support
Is there a way to use the GPU of an Apple Silcon Mac rather then using teh CPU ?
That woudl be much faster no ?

### PR DESCRIPTION
**Pull request by HeLehm**
This is kind of a dirty workaround but it does increase performance on mps devices.
As soon as PyTorch implements complex numbers for mps, this should be removed.

Fix #435 and fix #432

### CODE REVIEW

**Review by CarlGao4 — APPROVED** [review]
Have you tested that whether it can be used on mps?

**Comment by HeLehm** [comment]
Yes, I have and it does improve performance.

**Review by adefossez — APPROVED** [review]

**Comment by adefossez** [comment]
Seems like the linter is not passing, this is mandatory for accepting the PR!
Also, it is best to update the docs/release.md file, e.g. as in https://github.com/facebookresearch/demucs/pull/501/files#diff-87479ac9ea16cec5ceafce00b3b049d7d5920cb663afae42901f20e7c4e6705d

**Review by adefossez — CHANGES_REQUESTED** [review]
Linter is failing, update the docs/release.md file ! otherwise all good :)

File: demucs/hdemucs.py [hunk 1]
```diff
@@ -772,9 +772,21 @@ def forward(self, mix):
...
         x = x * std[:, None] + mean[:, None]
 
+        # to cpu as mps doesnt support complex numbers
+        # demucs issue #435 ##432
+        # NOTE: in this case z already is on cpu
+        # TODO: remove this when mps supports complex numbers
+        x_is_mps = x.device.type == "mps"
+        if x_is_mps:
+            x = x.cpu()
+
```
Comment (CarlGao4) [inline demucs/hdemucs.py:782]: Remove the empty spaces

File: demucs/htdemucs.py [hunk 2]
```diff
@@ -625,6 +625,14 @@ def forward(self, mix):
...
         x = x * std[:, None] + mean[:, None]
 
+        # to cpu as mps doesnt support complex numbers
+        # demucs issue #435 ##432
+        # NOTE: in this case z already is on cpu
+        # TODO: remove this when mps supports complex numbers
+        x_is_mps = x.device.type == "mps"
+        if x_is_mps:
+            x = x.cpu()
+
```
Comment (CarlGao4) [inline demucs/htdemucs.py:635]: Remove the empty spaces

**Review by CarlGao4 — COMMENTED** [review]
These are the changes required by linter

**Comment by HeLehm** [comment]
Thank you for your help! All requested changes are present in the last commit.

**Comment by CarlGao4** [comment]
I also have a suggestion:
Use `Fix #435 and fix #432` in your description so the issues will be automatically closed

**Comment by HeLehm** [comment]
> I also have a suggestion:
> Use `Fix #435 and fix #432` in your description so the issues will be automatically closed

Like this?

**Comment by CarlGao4** [comment]
Yes! Now you can see that these issues are linked with this PR:
image

**Comment by adefossez** [comment]
perfect!

**Comment by CarlGao4** [comment]
> do you know what is the speed up?

What do you mean?

**Comment by remorses** [comment]
How much faster is it on Mac now?

**Comment by HeLehm** [comment]
Yes I did some test (code in https://github.com/HeLehm/demucs/blob/mps_speed_test/mps_speed_test.ipynb)

Results:

Testing mdx
CPU time: 67.53s, MPS time: 224.81s
Speedup: 0.30x

Testing htdemucs
CPU time: 52.90s, MPS time: 14.44s
Speedup: 3.66x

Testing htdemucs_ft
CPU time: 210.97s, MPS time: 49.38s
Speedup: 4.27x

Testing htdemucs_6s
CPU time: 50.26s, MPS time: 12.76s
Speedup: 3.94x

Testing hdemucs_mmi
CPU time: 21.59s, MPS time: 62.02s
Speedup: 0.35x

Testing mdx_extra
CPU time: 79.57s, MPS time: 210.56s
Speedup: 0.38x

I think It is weird that these mdx models take longer on mps than on cpu... This should be investigated.

**Comment by CarlGao4** [comment]
More precisely, v3 (hdemucs) models are slower on mps
Can you try timing the cpu part when separating on mps? And see whether the rest parts are slower

**Comment by HeLehm** [comment]
> More precisely, v3 (hdemucs) models are slower on mps
> Can you try timing the cpu part when separating on mps? And see whether the rest parts are slower

Yes, will do. Sorry, I haven't found the time to to it yet. Will try next week.

**Comment by archive-r** [comment]
Great! This gives me a 6x speedup on my mbp14.

**Comment by HeLehm** [comment]
Ok I ran by tests more extensively by using this code:

```
def time_model(model_name, device, grad=False, n=1):
    model = pretrained.get_model(model_name)
    x = torch.randn(1, 2, 44100 * 30).to(device=device)  # 30 seconds of white noise for the test
    
    if grad:
        time = timeit.timeit(lambda: apply_model(model, x, device=device), number=n)
    else:
        with torch.no_grad():
            time = timeit.timeit(lambda: apply_model(model, x, device=device), number=n)

    del model
    del x
    gc.collect()

    time /= float(n)
    return time
```
Full Notebooke [here](https://github.com/HeLehm/demucs/blob/mps_speed_test/mps_speed_test.ipynb).
Here are the results:
**NOTE:** these numbers are not compatible to [my previous comment](https://github.com/facebookresearch/demucs/pull/503#issuecomment-1574858745), as I ran everything three times there and only once here.

Testing mdx with gradients
CPU time: 21.26s, MPS time: 11.79s
Speedup: 1.80x

Testing mdx without gradients
CPU time: 21.26s, MPS time: 13.40s
Speedup: 1.59x

Testing htdemucs with gradients
CPU time: 17.18s, MPS time: 6.94s
Speedup: 2.48x

Testing htdemucs without gradients
CPU time: 17.19s, MPS time: 4.23s
Speedup: 4.06x

Testing htdemucs_ft with gradients
CPU time: 68.47s, MPS time: 16.78s
Speedup: 4.08x

Testing htdemucs_ft without gradients
CPU time: 67.96s, MPS time: 16.69s
Speedup: 4.07x

Testing htdemucs_6s with gradients
CPU time: 17.09s, MPS time: 4.86s
Speedup: 3.52x

Testing htdemucs_6s without gradients
CPU time: 16.52s, MPS time: 4.28s
Speedup: 3.86x

Testing hdemucs_mmi with gradients
CPU time: 6.60s, MPS time: 5.83s
Speedup: 1.13x

Testing hdemucs_mmi without gradients
CPU time: 6.60s, MPS time: 3.85s
Speedup: 1.71x

Testing mdx_extra with gradients
CPU time: 26.35s, MPS time: 28.73s
Speedup: 0.92x

Testing mdx_extra without gradients
CPU time: 26.07s, MPS time: 22.78s
Speedup: 1.14x

As you can see, on CPU it doesn't really matter greatly if we use `torch.no_grad()` or not. I will try to do even more precise tests in the future.

It would also be helpful if someone else could run the notebook on their Apple M1/M2 device, to verify these results.

**Comment by CarlGao4** [comment]
Thanks a lot for your work!!!

What I asked and care is something like this:
```python
t1 = time.time()
...
t2 = time.time()
x.to("cpu")
...
x.to("mps")
t3 = time.time()
...
t4 = time.time()
print(t2 - t1 + t4 - t3)
```

To achieve this and pass `t2` and `t3` to your codes, I suggest that you can directly do this:
```python
# test.py
import __main__
__main__.start_time = time.time()
apply_model()
__main__.end_time = time.time()
print(__main__.end_time - __main__.start_time)
```
```python
# hdemucs.py / htdemucs.py
...
import __main__     # Use this to get the environment of your codes
st = time.time()    # before hdemucs.py#L779 and htdemucs.py#L632
x.cpu()
...
x.to("mps")
et = time.time()
__main__.start_time += et - st
```

**Comment by HeLehm** [comment]
Alright, I did. Code as always [here](https://github.com/HeLehm/demucs/tree/mps_speed_test).

Here is an image of the results:
On the y axis, we have the model and if grads (False -> run with torch.no_grad()).
On the x axis, we can see the different code blocks(look at [hdemucs.py](https://github.com/HeLehm/demucs/blob/mps_speed_test/demucs/hdemucs.py) / [htdemucs.py](https://github.com/HeLehm/demucs/blob/mps_speed_test/demucs/htdemucs.py) for more info)
You can Ignore the heat map scale on the right side, as each column is normalised independently.
The values are cpu - mps time, so big negative numbers mean mps was way faster.

The raw csv data is also uploaded [here](https://github.com/HeLehm/demucs/blob/mps_speed_test/demucs_performance.csv) and [here](https://github.com/HeLehm/demucs/blob/mps_speed_test/subtracted_data.csv).

demucs

**Comment by CarlGao4** [comment]
But, which operations must be done on cpu?

**Comment by CarlGao4** [comment]
What I can't understand is that why the results are different each time you test this

**Comment by MrSimmo** [comment]
Wonder if anyone knows what could be wrong - I've pulled the latest versions of the demucs folder from github and overwritten my local copy.

When I use -d mps on htdemucs or htdemucs_ft models - it works fine. Quite a bit quicker than without. But when I try using it on mdx or mdx_extra, I get a bunch of silent output files.

Using Apple M1 silicon on MacOS ventura, python 3.9.6.

Does anyone know what could be the issue?

**Comment by genevera** [comment]
It's still really, really slow running mdx models on mps.  I've got a M2Max with Sonoma and running HeLehm's code from the notebook gets me the following:

```bash
Testing htdemucs_6s with gradients
/Users/genevera/.pyenv/versions/miniconda3-latest/lib/python3.10/site-packages/demucs/apply.py:240: UserWarning: MPS: no support for int64 min/max ops, casting it to int32(Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/native/mps/operations/ReduceOps.mm:1271.)
  weight = (weight / weight.max())**transition_power
CPU time: 5.16s, MPS time: 11.04s
Speedup: 0.47x

Testing htdemucs_6s without gradients
CPU time: 5.17s, MPS time: 0.74s
Speedup: 6.98x

Testing hdemucs_mmi with gradients
CPU time: 2.37s, MPS time: 17.92s
Speedup: 0.13x

Testing hdemucs_mmi without gradients
CPU time: 2.37s, MPS time: 22.23s
Speedup: 0.11x

Testing htdemucs_ft with gradients
CPU time: 34.05s, MPS time: 6.35s
Speedup: 5.36x

Testing htdemucs_ft without gradients
CPU time: 22.01s, MPS time: 3.11s
Speedup: 7.09x

Testing htdemucs with gradients
CPU time: 5.61s, MPS time: 0.78s
Speedup: 7.15x

Testing htdemucs without gradients
CPU time: 5.65s, MPS time: 0.80s
Speedup: 7.04x

Testing repro_mdx_a with gradients
CPU time: 7.51s, MPS time: 79.45s
Speedup: 0.09x

Testing repro_mdx_a without gradients
CPU time: 7.32s, MPS time: 100.32s
Speedup: 0.07x

Testing mdx_extra_q with gradients
CPU time: 8.64s, MPS time: 65.12s
Speedup: 0.13x

Testing mdx_extra_q without gradients
CPU time: 8.87s, MPS time: 67.84s
Speedup: 0.13x

Testing mdx with gradients
CPU time: 6.65s, MPS time: 61.50s
Speedup: 0.11x

Testing mdx without gradients
CPU time: 6.93s, MPS time: 65.82s
Speedup: 0.11x

Testing repro_mdx_a_time_only with gradients
CPU time: 6.01s, MPS time: 65.09s
Speedup: 0.09x

Testing repro_mdx_a_time_only without gradients
CPU time: 5.82s, MPS time: 23.45s
Speedup: 0.25x

Testing mdx_extra with gradients
CPU time: 9.19s, MPS time: 112.45s
Speedup: 0.08x

Testing mdx_extra without gradients
CPU time: 9.43s, MPS time: 77.31s
Speedup: 0.12x

Testing repro_mdx_a_hybrid_only with gradients
CPU time: 9.16s, MPS time: 90.22s
Speedup: 0.10x

Testing repro_mdx_a_hybrid_only without gradients
CPU time: 9.03s, MPS time: 100.25s
Speedup: 0.09x

Testing mdx_q with gradients
CPU time: 7.03s, MPS time: 117.97s
Speedup: 0.06x

Testing mdx_q without gradients
CPU time: 6.82s, MPS time: 64.27s
Speedup: 0.11x
```

**Comment by Andy671** [comment]
Thanks so much for this PR. You saved us hours of processing time!
